### PR TITLE
Add graph capture and shaders to improve DirectML performance

### PR DESCRIPTION
**Pull request by PatriceVignola**
This adds graph capture functionality and shaders to the DML EP. Note that most files under the `src/models/dml` folder are copied from onnxruntime and don't need to get reviewed.

### CODE REVIEW

File: src/python/py/models/builder.py [hunk 1]
```diff
@@ -251,6 +254,9 @@ def make_genai_config(self, model_name_or_path, extra_kwargs, out_dir):
         with open(os.path.join(out_dir,"genai_config.json"), "w") as f:
             json.dump(genai_config, f, indent=4)
 
+    def cuda_graph_enabled(self):
+        return self.ep_attrs["cuda"]["enable_cuda_graph"] == '1' or self.ep == "dml"
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:258]: I think this function can be removed since it is not used anywhere.

File: src/python/py/models/builder.py [hunk 2]
```diff
@@ -171,8 +171,11 @@ def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options):
             self.attention_attrs["use_packed_matmul"] = self.ep != "dml" and self.num_attn_heads == self.num_kv_heads
 
             # GQA + Rot.Emb. does not require `position ids` as input
-            self.attention_attrs["use_rotemb_in_attn"] = True
-            self.input_names.remove("position_ids")
+            if self.ep == "cuda":
+                self.attention_attrs["use_rotemb_in_attn"] = True
+                self.input_names.remove("position_ids")
+
+        self.past_present_share_buffer = self.attention_attrs["op_type"] == "GroupQueryAttention"
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:178]: There appears to be some duplicate changes after merging branches. I think you can remove this line since it's defined again below.

File: src/models/input_ids.h [hunk 3]
```diff
@@ -23,7 +23,13 @@ struct InputIDs {
   std::unique_ptr<OrtValue> value_;
 
   // Used for decoding runs with cuda graphs.
-  std::unique_ptr<StaticBuffer> sb_input_ids_;
+  StaticBuffer* sb_input_ids_ = nullptr;
```
Comment (RyanUnderhill) [inline src/models/input_ids.h:26]:
```suggestion
  StaticBuffer* sb_input_ids_{};
```

File: src/models/input_ids.h [hunk 4]
```diff
@@ -23,7 +23,13 @@ struct InputIDs {
...
+
+#if USE_DML
+  std::unique_ptr<OrtValue> value_int32_;
+  StaticBuffer* sb_input_ids_int32_ = nullptr;
```
Comment (RyanUnderhill) [inline src/models/input_ids.h:30]:
```suggestion
  StaticBuffer* sb_input_ids_int32_{};
```

File: src/models/logits.cpp [hunk 5]
```diff
@@ -9,33 +9,40 @@ Logits::Logits(const Model& model, State& state)
...
   size_t element_count = shape_[0] * shape_[1] * shape_[2];
-  // DML doesn't support on-device scoring yet, so fall back to the CPU
-  auto& allocator = model_.device_type_ == DeviceType::DML ? model_.allocator_cpu_ : *model_.allocator_device_;
 
   // Convert from float16 to float32 if necessary
-  if (type_ == Ort::TypeToTensorType<Ort::Float16_t>::type)
-    ConvertFp16ToFp32(allocator, *value16_, value32_, model_.device_type_, model_.cuda_stream_);
+  if (type_ == Ort::TypeToTensorType<Ort::Float16_t>::type) {
+#if USE_DML
+    DmlHelpers::DmlCastInputToOutput(
```
Comment (RyanUnderhill) [inline src/models/logits.cpp:34]: Could this be added to ConvertFp16ToFp32, it already works for CPU & CUDA, DML shouldn't be a special case here.

File: src/models/logits.cpp [hunk 6]
```diff
@@ -63,16 +74,49 @@ RoamingArray<float> Logits::Get() {
...
-          CudaCheck() == cudaMemcpyAsync(target.data(), source.data(), source.size_bytes(), cudaMemcpyDeviceToDevice, state_.params_->cuda_stream);
-        else
+          case DeviceType::CUDA: {
+            auto logits = std::span<float>{value32_->GetTensorMutableData<float>(), element_count};
+            auto logits_next = gpu_span<float>{value_next->GetTensorMutableData<float>(), element_count};
+            auto target = logits_next.subspan(vocab_index, vocab_size);
+            std::span<const float> source = logits.subspan(vocab_index * seq_length + token_index * vocab_size, vocab_size);
+            CudaCheck() == cudaMemcpyAsync(target.data(), source.data(), source.size_bytes(), cudaMemcpyDeviceToDevice, state_.params_->cuda_stream);
+
+          } break;
```
Comment (RyanUnderhill) [inline src/models/logits.cpp:110]:
```suggestion
          case DeviceType::CPU:
          case DeviceType::CUDA: {
            auto logits = std::span<float>{value32_->GetTensorMutableData<float>(), element_count};
            auto logits_next = gpu_span<float>{value_next->GetTensorMutableData<float>(), element_count};
            auto target = logits_next.subspan(vocab_index, vocab_size);
            std::span<const float> source = logits.subspan(vocab_index * seq_length + token_index * vocab_size, vocab_size);
            if(model_.device_type_ == DeviceType::CUDA)
#if USE_CUDA            
            CudaCheck() == cudaMemcpyAsync(target.data(), source.data(), source.size_bytes(), cudaMemcpyDeviceToDevice, state_.params_->cuda_stream);
#else
           throw std::runtime_error("Unexpected CUDA device usage");
#endif
         else
             copy(source, target);
          } break;
```

Comment (RyanUnderhill) [reply src/models/logits.cpp:110]: Code suggestion above is to remove the duplicate code

File: src/models/logits.h [hunk 7]
```diff
@@ -21,8 +21,13 @@ struct Logits {
   std::unique_ptr<OrtValue> value16_;  // When model output is fp16
 
   // Used for decoding runs with cuda graphs.
-  std::unique_ptr<StaticBuffer> sb_logits32_;
-  std::unique_ptr<StaticBuffer> sb_logits16_;
+  StaticBuffer* sb_logits32_ = nullptr;
+  StaticBuffer* sb_logits16_ = nullptr;
```
Comment (RyanUnderhill) [inline src/models/logits.h:25]:
```suggestion
  StaticBuffer* sb_logits32_{};
  StaticBuffer* sb_logits16_{};
```

File: src/models/model.cpp [hunk 8]
```diff
@@ -9,9 +9,9 @@
 #include "whisper.h"
 #include "kernels.h"
 #ifdef USE_DML
-//  Because dml_provider_factory includes windows headers that #define min and max, this next line will prevent this from happening
-#define NOMINMAX
+#include <wil/wrl.h>
```
Comment (RyanUnderhill) [inline src/models/model.cpp:12]: How did you get past the min and max macros?

File: src/models/position_inputs.h [hunk 9]
```diff
@@ -46,11 +51,18 @@ struct PositionInputs {
   std::vector<int32_t> initial_sequence_lengths_;
 
   // Used for decoding runs with cuda graphs.
-  std::unique_ptr<StaticBuffer> sb_position_ids_;
-  std::unique_ptr<StaticBuffer> sb_attention_mask_;
+  StaticBuffer* sb_position_ids_ = nullptr;
+  StaticBuffer* sb_attention_mask_ = nullptr;
```
Comment (RyanUnderhill) [inline src/models/position_inputs.h:55]:
```suggestion
  StaticBuffer* sb_position_ids_{};
  StaticBuffer* sb_attention_mask_{};
```

File: src/models/position_inputs.h [hunk 10]
```diff
@@ -46,11 +51,18 @@ struct PositionInputs {
...
 
   bool is_first_posid_update_{true};
   bool is_first_mask_update_{true};
+
+#ifdef USE_DML
+  std::optional<DmlUpdateMaskKernel> dml_update_mask_kernel_;
+  StaticBuffer* sb_attention_mask_next_ = nullptr;
+  std::optional<DmlIncrementValuesKernel> dml_update_position_ids_kernel_;
+  bool is_second_mask_update_{false};
```
Comment (RyanUnderhill) [inline src/models/position_inputs.h:64]:
```suggestion
  StaticBuffer* sb_attention_mask_next_{};
  std::optional<DmlIncrementValuesKernel> dml_update_position_ids_kernel_;
  bool is_second_mask_update_{};
```

File: src/models/dml/dml_command_allocator_ring.h [hunk 11]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (RyanUnderhill) [inline src/models/dml/dml_command_allocator_ring.h:1]: Can the dml specific files be under the src directory?
Since there'll be dml code used by the non model code also.

File: src/models/dml/dml_command_queue.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  };
+
+  std::deque<QueuedReference> queued_references_;
+
+  ComPtr<ID3D12CommandQueue> queue_;
+  D3D12_COMMAND_LIST_TYPE type_;
+
+  ComPtr<ID3D12Fence> fence_;
+  uint64_t last_fence_value_ = 0;
+  bool closing_ = false;
```
Comment (RyanUnderhill) [inline src/models/dml/dml_command_queue.h:53]: I didn't review the DML specific files carefully, are they copies of another codebase?

File: src/models/model.cpp (hunk 8 above)
Comment (PatriceVignola) [reply src/models/model.cpp:12]: We do it in the cmake files instead. It should be cleaner.

File: src/models/dml/dml_command_queue.h (hunk 12 above)
Comment (PatriceVignola) [reply src/models/dml/dml_command_queue.h:53]: Mostly from onnxruntime, but with changes related to this repo due to different dependencies (e.g. no GSL here, and the files in ORT repo were depending on some ort-specific objects that I removed). With more time the goal is to move those over to a different repo that can be consumed by both ORT, ort-genai and any other frameworks that needs it.

**Review by RyanUnderhill — APPROVED** [review]

**Review by kunal-vaishnavi — APPROVED** [review]
